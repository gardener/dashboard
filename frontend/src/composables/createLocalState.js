//
// SPDX-FileCopyrightText: 2023 SAP SE or an SAP affiliate company and Gardener contributors
//
// SPDX-License-Identifier: Apache-2.0
//

import { effectScope } from 'vue'

export function createLocalState (stateFactory) {
  let state
  const scope = effectScope(true)

  return (...args) => {
    state = scope.run(() => stateFactory(...args))
    return state
  }
}
