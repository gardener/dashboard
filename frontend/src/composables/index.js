//
// SPDX-FileCopyrightText: 2023 SAP SE or an SAP affiliate company and Gardener contributors
//
// SPDX-License-Identifier: Apache-2.0
//

export { useApi } from './useApi'
export { useAuthn } from './useAuthn'
export { useLogger } from './useLogger'
export { useNav } from './useNav'
export { useTheme } from './useTheme'
export { useToken } from './useToken'
export { useUser } from './useUser'
