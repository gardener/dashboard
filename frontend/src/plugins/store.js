//
// SPDX-FileCopyrightText: 2023 SAP SE or an SAP affiliate company and Gardener contributors
//
// SPDX-License-Identifier: Apache-2.0
//

import pinia from '@/store'

export default {
  install (app) {
    app.use(pinia)
  },
}
