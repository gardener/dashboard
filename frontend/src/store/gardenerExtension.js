//
// SPDX-FileCopyrightText: 2023 SAP SE or an SAP affiliate company and Gardener contributors
//
// SPDX-License-Identifier: Apache-2.0
//

import {
  defineStore,
  acceptHMRUpdate,
} from 'pinia'
import {
  ref,
  computed,
} from 'vue'

import { useApi } from '@/composables/useApi'

import {
  map,
  flatMap,
  filter,
  get,
  some,
  find,
  sortBy,
} from '@/lodash'

export const useGardenerExtensionStore = defineStore('gardenerExtension', () => {
  const api = useApi()

  const list = ref(null)

  const isInitial = computed(() => {
    return list.value === null
  })

  const gardenerExtensionList = computed(() => {
    return list.value
  })

  async function fetchGardenerExtensions () {
    const response = await api.getGardenerExtensions()
    list.value = response.data
  }

  const sortedDnsProviderList = computed(() => {
    const supportedProviderTypes = ['aws-route53', 'azure-dns', 'azure-private-dns', 'google-clouddns', 'openstack-designate', 'alicloud-dns', 'infoblox-dns', 'netlify-dns']
    const resources = flatMap(list.value, 'resources')
    const dnsProvidersFromDnsRecords = filter(resources, ['kind', 'DNSRecord'])

    const dnsProviderList = map(supportedProviderTypes, type => {
      const dnsProvider = find(dnsProvidersFromDnsRecords, ['type', type])
      return {
        type,
        primary: get(dnsProvider, 'primary', false),
      }
    })

    const dnsServiceExtensionDeployed = some(list.value, ['name', 'extension-shoot-dns-service'])
    return dnsServiceExtensionDeployed
      ? dnsProviderList
      : filter(dnsProviderList, 'primary') // return only primary DNS Providers backed by DNSRecord
  })

  const dnsProviderTypes = computed(() => {
    return map(sortedDnsProviderList.value, 'type')
  })

  const dnsProviderTypesWithPrimarySupport = computed(() => {
    return map(filter(sortedDnsProviderList.value, 'primary'), 'type')
  })

  const networkingTypes = computed(() => {
    const resources = flatMap(list.value, 'resources')
    const networkingResources = filter(resources, ['kind', 'Network'])
    return map(networkingResources, 'type')
  })

  const networkingTypeList = computed(() => {
    return sortBy(networkingTypes.value)
  })

  return {
    list,
    isInitial,
    gardenerExtensionList,
    fetchGardenerExtensions,
    dnsProviderTypes,
    dnsProviderTypesWithPrimarySupport,
    networkingTypes,
    networkingTypeList,
  }
})

if (import.meta.hot) {
  import.meta.hot.accept(acceptHMRUpdate(useGardenerExtensionStore, import.meta.hot))
}
