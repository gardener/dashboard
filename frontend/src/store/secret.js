//
// SPDX-FileCopyrightText: 2023 SAP SE or an SAP affiliate company and Gardener contributors
//
// SPDX-License-Identifier: Apache-2.0
//

import {
  defineStore,
  acceptHMRUpdate,
} from 'pinia'
import {
  ref,
  computed,
} from 'vue'

import { useApi } from '@/composables/useApi'

import { isOwnSecret } from '@/utils'

import { useAuthzStore } from './authz'
import { useAppStore } from './app'

import {
  findIndex,
  find,
  filter,
  matches,
} from '@/lodash'

function eqlNameAndNamespace ({ namespace, name }) {
  return matches({ metadata: { namespace, name } })
}

export const useSecretStore = defineStore('secret', () => {
  const api = useApi()
  const appStore = useAppStore()
  const authzStore = useAuthzStore()

  const list = ref(null)

  const isInitial = computed(() => {
    return list.value === null
  })

  function $reset () {
    list.value = null
  }

  async function fetchSecrets () {
    const namespace = authzStore.namespace
    try {
      const response = await api.getCloudProviderSecrets({ namespace })
      list.value = response.data
    } catch (err) {
      $reset()
      throw err
    }
  }

  function getSecret ({ namespace, name }) {
    return find(list.value, eqlNameAndNamespace({ namespace, name }))
  }

  async function createSecret (obj) {
    const {
      namespace = authzStore.namespace,
    } = obj.metadata
    const response = await api.createCloudProviderSecret({ namespace, data: obj })

    appStore.setSuccess('Cloud Provider secret created')
    replace(response.data)
  }

  async function updateSecret (obj) {
    const {
      namespace = authzStore.namespace,
      name,
    } = obj.metadata
    const response = await api.updateCloudProviderSecret({ namespace, name, data: obj })
    appStore.setSuccess('Cloud Provider secret updated')
    replace(response.data)
  }

  async function deleteSecret (name) {
    const namespace = authzStore.namespace
    const response = await api.deleteCloudProviderSecret({ namespace, name })
    appStore.setSuccess('Cloud Provider secret deleted')
    remove(response.data)
  }

  const infrastructureSecretList = computed(() => {
    return filter(list.value, secret => {
      return !!secret.metadata.cloudProviderKind
    })
  })

  const dnsSecretList = computed(() => {
    return filter(list.value, secret => {
      return !!secret.metadata.dnsProviderName && isOwnSecret(secret) // secret binding not supported
    })
  })

  function getCloudProviderSecretByName ({ namespace, name }) {
    return find(list.value, eqlNameAndNamespace({ name, namespace }))
  }

  function infrastructureSecretsByCloudProfileName (cloudProfileName) {
    return filter(list.value, ['metadata.cloudProfileName', cloudProfileName])
  }

  function dnsSecretsByProviderKind (dnsProviderName) {
    return filter(list.value, secret => {
      return secret.metadata.dnsProviderName === dnsProviderName && isOwnSecret(secret) // secret binding not supported
    })
  }

  function replace (obj) {
    const index = findIndex(list.value, eqlNameAndNamespace(obj.metadata))
    if (index !== -1) {
      list.value.splice(index, 1, {
        ...list.value[index],
        ...obj,
      })
    } else {
      list.value.push(obj)
    }
  }

  function remove (obj) {
    const index = findIndex(list.value, eqlNameAndNamespace(obj.metadata))
    if (index !== -1) {
      list.value.splice(index, 1)
    }
  }

  return {
    list,
    isInitial,
    fetchSecrets,
    getSecret,
    updateSecret,
    createSecret,
    deleteSecret,
    infrastructureSecretList,
    dnsSecretList,
    dnsSecretsByProviderKind,
    getCloudProviderSecretByName,
    infrastructureSecretsByCloudProfileName,
    $reset,
  }
})

if (import.meta.hot) {
  import.meta.hot.accept(acceptHMRUpdate(useSecretStore, import.meta.hot))
}
