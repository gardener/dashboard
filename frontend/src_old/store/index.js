//
// SPDX-FileCopyrightText: 2021 SAP SE or an SAP affiliate company and Gardener contributors
//
// SPDX-License-Identifier: Apache-2.0
//

import { listProjectTerminalShortcuts } from '@/utils/api'

// actions
export const actions = {
  fetchCloudProfiles ({ dispatch }) {
    return dispatch('cloudProfiles/getAll')
      .catch(err => {
        dispatch('setError', err)
      })
  },
  async fetchGardenerExtensions ({ dispatch }) {
    try {
      await dispatch('gardenerExtensions/getAll')
    } catch (err) {
      dispatch('setError', err)
    }
  },
  async fetchSeeds ({ dispatch }) {
    try {
      await dispatch('seeds/getAll')
    } catch (err) {
      dispatch('setError', err)
    }
  },
  fetchProjects ({ dispatch }) {
    return dispatch('projects/getAll')
      .catch(err => {
        dispatch('setError', err)
      })
  },
  fetchMembers ({ dispatch, commit }) {
    return dispatch('members/getAll')
      .catch(err => {
        dispatch('setError', err)
      })
  },
  fetchcloudProviderSecrets ({ dispatch, commit }) {
    return dispatch('cloudProviderSecrets/getAll')
      .catch(err => {
        dispatch('setError', err)
      })
  },
  getShootInfo ({ dispatch, commit }, { name, namespace }) {
    return dispatch('shoots/getInfo', { name, namespace })
      .catch(err => {
        dispatch('setError', err)
      })
  },
  setSelectedShoot ({ dispatch }, metadata) {
    return dispatch('shoots/setSelection', metadata)
      .catch(err => {
        dispatch('setError', err)
      })
  },
  async setShootListFilters ({ dispatch, getters }, value) {
    try {
      await dispatch('shoots/setShootListFilters', value)
    } catch (err) {
      dispatch('setError', err)
    }
  },
  async setShootListFilter ({ dispatch, getters }, { filter, value }) {
    try {
      await dispatch('shoots/setShootListFilter', { filter, value })
    } catch (err) {
      dispatch('setError', err)
    }
  },
  createProject ({ dispatch, commit }, data) {
    return dispatch('projects/create', data)
      .then(res => {
        dispatch('setAlert', { message: 'Project created', type: 'success' })
        return res
      })
  },
  patchProject ({ dispatch, commit }, data) {
    return dispatch('projects/patch', data)
  },
  updateProject ({ dispatch, commit }, data) {
    return dispatch('projects/update', data)
      .then(res => {
        dispatch('setAlert', { message: 'Project updated', type: 'success' })
        return res
      })
  },
  deleteProject ({ dispatch, commit }, data) {
    return dispatch('projects/delete', data)
      .then(res => {
        dispatch('setAlert', { message: 'Project deleted', type: 'success' })
        return res
      })
  },
  createCloudProviderSecret ({ dispatch, commit }, data) {
    return dispatch('cloudProviderSecrets/create', data)
      .then(res => {
        dispatch('setAlert', { message: 'Cloud Provider secret created', type: 'success' })
        return res
      })
  },
  updateCloudProviderSecret ({ dispatch, commit }, data) {
    return dispatch('cloudProviderSecrets/update', data)
      .then(res => {
        dispatch('setAlert', { message: 'Cloud Provider secret updated', type: 'success' })
        return res
      })
  },
  deleteCloudProviderSecret ({ dispatch, commit }, data) {
    return dispatch('cloudProviderSecrets/delete', data)
      .then(res => {
        dispatch('setAlert', { message: 'Cloud Provider secret deleted', type: 'success' })
        return res
      })
  },
  createShoot ({ dispatch, commit }, data) {
    return dispatch('shoots/create', data)
      .then(res => {
        dispatch('setAlert', { message: 'Cluster created', type: 'success' })
        return res
      })
  },
  deleteShoot ({ dispatch, commit }, { name, namespace }) {
    return dispatch('shoots/delete', { name, namespace })
      .then(res => {
        dispatch('setAlert', { message: 'Cluster marked for deletion', type: 'success' })
        return res
      })
  },
  subscribeShoots ({ dispatch, commit }) {
    (async () => {
      try {
        await dispatch('shoots/subscribe')
      } catch (err) {
        commit('SET_ALERT', { message: err.message, type: 'error' })
      }
    })()
  },
  unsubscribeShoots ({ dispatch, commit }) {
    (async () => {
      try {
        await dispatch('shoots/unsubscribe')
      } catch (err) {
        commit('SET_ALERT', { message: err.message, type: 'error' })
      }
    })()
  },
  async addMember ({ dispatch, commit }, payload) {
    const result = await dispatch('members/add', payload)
    await dispatch('setAlert', { message: 'Member added', type: 'success' })
    return result
  },
  async updateMember ({ dispatch, commit }, payload) {
    const result = await dispatch('members/update', payload)
    await dispatch('setAlert', { message: 'Member updated', type: 'success' })
    return result
  },
  async deleteMember ({ dispatch, commit }, payload) {
    try {
      const result = await dispatch('members/delete', payload)
      await dispatch('setAlert', { message: 'Member deleted', type: 'success' })
      return result
    } catch (err) {
      await dispatch('setError', { message: `Delete member failed. ${err.message}` })
    }
  },
  async resetServiceAccount ({ dispatch, commit }, payload) {
    try {
      const result = await dispatch('members/resetServiceAccount', payload)
      await dispatch('setAlert', { message: 'Service Account Reset', type: 'success' })
      return result
    } catch (err) {
      await dispatch('setError', { message: `Failed to Reset Service Account ${err.message}` })
    }
  },
  async setNamespace ({ dispatch, commit, state }, namespace) {
    if (namespace && state.namespace !== namespace) {
      commit('SET_NAMESPACE', namespace)
      await dispatch('refreshSubjectRules', namespace)
    }
    return state.namespace
  },
  async ensureProjectTerminalShortcutsLoaded ({ commit, dispatch, state }) {
    const { namespace, projectTerminalShortcuts } = state
    if (!projectTerminalShortcuts || projectTerminalShortcuts.namespace !== namespace) {
      try {
        const { data: items } = await listProjectTerminalShortcuts({ namespace })
        commit('SET_PROJECT_TERMINAL_SHORTCUTS', {
          namespace,
          items,
        })
      } catch (err) {
        // eslint-disable-next-line no-console
        console.warn('Failed to list project terminal shortcuts:', err.message)
      }
    }
  },
}
