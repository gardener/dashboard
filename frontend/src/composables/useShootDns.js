//
// SPDX-FileCopyrightText: 2024 SAP SE or an SAP affiliate company and Gardener contributors
//
// SPDX-License-Identifier: Apache-2.0
//

import { computed } from 'vue'

import { useGardenerExtensionStore } from '@/store/gardenerExtension'
import { useSecretStore } from '@/store/secret'

import { useShootResources } from '@/composables/useShootResources'
import { useShootExtensions } from '@/composables/useShootExtensions'

import {
  get,
  set,
  unset,
  find,
  map,
  head,
  includes,
  filter,
  isEmpty,
} from '@/lodash'

export const useShootDns = (manifest, options) => {
  const {
    gardenerExtensionStore = useGardenerExtensionStore(),
    secretStore = useSecretStore(),
  } = options

  /* resources */
  const {
    resources,
    deleteResource,
    setResource,
    getResourceRefName,
  } = useShootResources(manifest)

  /* extensions */
  const {
    extensions,
    setExtension,
    deleteExtension,
  } = useShootExtensions(manifest)

  const dnsServiceExtension = computed(() => {
    return find(extensions.value, ['type', 'shoot-dns-service'])
  })

  const dnsServiceExtensionProviders = computed({
    get () {
      return get(dnsServiceExtension.value, 'providerConfig.providers', [])
    },
    set (value) {
      set(dnsServiceExtension.value, 'providerConfig.providers', value)
    },
  })

  /* dns */
  const dnsDomain = computed({
    get () {
      return get(manifest.value, 'spec.dns.domain')
    },
    set (value) {
      set(manifest.value, 'spec.dns.domain', value)
    },
  })

  const dnsProviders = computed({
    get () {
      return get(manifest.value, 'spec.dns.providers')
    },
    set (value) {
      if (value) {
        set(manifest.value, 'spec.dns.providers', value)
      } else {
        unset(manifest.value, 'spec.dns.providers')
      }
    },
  })

  function resetDnsPrimaryProvider () {
    if (!dnsDomain.value) {
      dnsProviders.value = undefined
    }
  }

  const dnsPrimaryProvider = computed(() => {
    return find(dnsProviders.value, ['primary', true])
  })

  function patchDnsPrimaryProvider (data) {
    // 'spec.dns.providers' may only contain a single primary provider. For historical reasons it is still an array
    dnsProviders.value = [{
      primary: true,
      ...dnsPrimaryProvider.value,
      ...data,
    }]
  }

  const dnsPrimaryProviderType = computed({
    get () {
      return dnsPrimaryProvider.value?.type
    },
    set (type) {
      patchDnsPrimaryProvider({ type })
    },
  })

  const dnsPrimaryProviderSecretName = computed({
    get () {
      return dnsPrimaryProvider.value?.secretName
    },
    set (secretName) {
      patchDnsPrimaryProvider({ secretName })
    },
  })

  const primaryDnsServiceExtensionProvider = computed(() => {
    if (!dnsPrimaryProvider.value) {
      return undefined
    }
    return find(dnsServiceExtensionProviders.value, provider => {
      const providerSecretName = getResourceRefName(provider.secretName)
      return provider.type === dnsPrimaryProvider.value.type &&
        providerSecretName === dnsPrimaryProvider.value.secretName
    })
  })

  const hasDnsServiceExtensionProviderForCustomDomain = computed(() => {
    const provider = primaryDnsServiceExtensionProvider.value
    return provider
      ? includes(provider.domains?.include, dnsDomain.value)
      : false
  })

  function getDnsServiceExtensionResourceName (secretName) {
    return secretName
      ? `shoot-dns-service-${secretName}`
      : undefined
  }

  function addDnsServiceExtensionProviderForCustomDomain () {
    if (hasDnsServiceExtensionProviderForCustomDomain.value) {
      return
    }
    const customDomainProvider = primaryDnsServiceExtensionProvider.value ?? addDnsServiceExtensionProvider(dnsPrimaryProvider.value)
    if (isEmpty(customDomainProvider.domains?.include)) {
      set(customDomainProvider, 'domains.include', [dnsDomain.value])
    } else {
      customDomainProvider.domains.include.push(dnsDomain.value)
    }
  }

  function getDefaultSecretName (type) {
    const secrets = secretStore.dnsSecretsByProviderKind(type)
    // find unused secret
    const usedResourceNames = map(resources.value, 'name')
    const secret = find(secrets, secret => {
      const resourceName = getDnsServiceExtensionResourceName(secret.metadata.name)
      return !includes(usedResourceNames, resourceName)
    })
    return get(secret, 'metadata.secretRef.name')
  }

  function addDnsServiceExtensionProvider (options = {}) {
    const {
      type = head(gardenerExtensionStore.dnsProviderTypes),
      secretName = getDefaultSecretName(type),
    } = options

    const resourceName = getDnsServiceExtensionResourceName(secretName)
    const provider = {
      type,
      secretName: resourceName, // resourceName is the secret name
    }
    if (isEmpty(dnsServiceExtensionProviders.value)) {
      setExtension({
        type: 'shoot-dns-service',
        providerConfig: {
          apiVersion: 'service.dns.extensions.gardener.cloud/v1alpha1',
          kind: 'DNSConfig',
          syncProvidersFromShootSpecDNS: false,
          providers: [provider],
        },
      })
    } else {
      dnsServiceExtensionProviders.value.push(provider)
    }

    setResource({
      name: resourceName,
      resourceRef: {
        apiVersion: 'v1',
        kind: 'Secret',
        name: secretName,
      },
    })

    return provider
  }

  function deleteDnsServiceExtensionProvider (index) {
    const provider = dnsServiceExtensionProviders.value[index]
    if (!provider) {
      return
    }

    const resourceName = provider.secretName
    deleteResource(resourceName)

    dnsServiceExtensionProviders.value.splice(index, 1)
    if (isEmpty(dnsServiceExtensionProviders.value)) {
      deleteExtension('shoot-dns-service')
    }
  }

  function forceMigrateSyncDnsProvidersToFalse () {
    if (get(dnsServiceExtension.value, 'providerConfig.syncProvidersFromShootSpecDNS') === true) {
      // Migrate from old DNS configuration to new DNS configuration
      set(dnsServiceExtension.value, 'providerConfig.syncProvidersFromShootSpecDNS', false)
      dnsProviders.value = filter(dnsProviders.value, ['primary', true])
    }
  }

  return {
    /* dns */
    dnsDomain,
    dnsPrimaryProviderType,
    dnsPrimaryProviderSecretName,
    resetDnsPrimaryProvider,
    forceMigrateSyncDnsProvidersToFalse,
    dnsServiceExtensionProviders,
    addDnsServiceExtensionProvider,
    deleteDnsServiceExtensionProvider,
    hasDnsServiceExtensionProviderForCustomDomain,
    addDnsServiceExtensionProviderForCustomDomain,
    getDnsServiceExtensionResourceName,
    deleteResource,
    setResource,
    getResourceRefName,
  }
}
