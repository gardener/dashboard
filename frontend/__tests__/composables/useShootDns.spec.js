//
// SPDX-FileCopyrightText: 2021 SAP SE or an SAP affiliate company and Gardener contributors
//
// SPDX-License-Identifier: Apache-2.0
//
import {
  reactive,
  toRef,
} from 'vue'
import {
  setActivePinia,
  createPinia,
} from 'pinia'

import { useGardenerExtensionStore } from '@/store/gardenerExtension'
import { useSecretStore } from '@/store/secret'

import { useShootDns } from '@/composables/useShootDns'

describe('composables', () => {
  describe('useShootDns', () => {
    const manifest = reactive({})
    let shootDns

    beforeEach(() => {
      setActivePinia(createPinia())
      manifest.spec = {}
      const secretStore = useSecretStore()
      secretStore.list = global.fixtures.secrets
      const gardenerExtensionStore = useGardenerExtensionStore()
      gardenerExtensionStore.list = global.fixtures.gardenerExtensions

      const composable = useShootDns(toRef(manifest), {
        gardenerExtensionStore,
        secretStore,
      })
      shootDns = reactive(composable)
    })

    it('should add extension dns providers', () => {
      shootDns.addDnsServiceExtensionProvider()
      shootDns.dnsDomain = 'example.org'

      expect(shootDns.dnsServiceExtensionProviders).toHaveLength(1)

      expect(manifest.spec).toMatchSnapshot()
    })

    it('should delete extension dns providers', () => {
      shootDns.addDnsServiceExtensionProvider()
      shootDns.addDnsServiceExtensionProvider()

      expect(shootDns.dnsServiceExtensionProviders).toHaveLength(2)
      expect(manifest.spec).toMatchSnapshot('two providers added')

      shootDns.deleteDnsServiceExtensionProvider(0)
      expect(manifest.spec).toMatchSnapshot('one provider deleted')

      // Delete last extension dns provider
      shootDns.deleteDnsServiceExtensionProvider(0)
      expect(manifest.spec).toMatchSnapshot('last provider deleted')
    })

    it('should add primary dns provider', () => {
      shootDns.dnsDomain = 'example.org'
      shootDns.dnsPrimaryProviderType = 'foo'
      shootDns.dnsPrimaryProviderSecretName = 'bar'

      expect(manifest.spec).toMatchSnapshot()
    })

    it('should add extension custom domain dns provider', () => {
      shootDns.dnsDomain = 'example.org'
      shootDns.dnsPrimaryProviderType = 'foo'
      shootDns.dnsPrimaryProviderSecretName = 'bar'

      expect(shootDns.hasDnsServiceExtensionProviderForCustomDomain).toBe(false)

      shootDns.addDnsServiceExtensionProviderForCustomDomain()
      expect(shootDns.hasDnsServiceExtensionProviderForCustomDomain).toBe(true)

      expect(manifest.spec).toMatchSnapshot()
    })
  })
})
