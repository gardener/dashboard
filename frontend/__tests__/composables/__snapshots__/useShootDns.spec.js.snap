// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`composables > useShootDns > should add extension custom domain dns provider 1`] = `
{
  "dns": {
    "domain": "example.org",
    "providers": [
      {
        "primary": true,
        "secretName": "bar",
        "type": "foo",
      },
    ],
  },
  "extensions": [
    {
      "providerConfig": {
        "apiVersion": "service.dns.extensions.gardener.cloud/v1alpha1",
        "kind": "DNSConfig",
        "providers": [
          {
            "domains": {
              "include": [
                "example.org",
              ],
            },
            "secretName": "shoot-dns-service-bar",
            "type": "foo",
          },
        ],
        "syncProvidersFromShootSpecDNS": false,
      },
      "type": "shoot-dns-service",
    },
  ],
  "resources": [
    {
      "name": "shoot-dns-service-bar",
      "resourceRef": {
        "apiVersion": "v1",
        "kind": "Secret",
        "name": "bar",
      },
    },
  ],
}
`;

exports[`composables > useShootDns > should add extension dns providers 1`] = `
{
  "dns": {
    "domain": "example.org",
  },
  "extensions": [
    {
      "providerConfig": {
        "apiVersion": "service.dns.extensions.gardener.cloud/v1alpha1",
        "kind": "DNSConfig",
        "providers": [
          {
            "secretName": "shoot-dns-service-aws-route53-secret",
            "type": "aws-route53",
          },
        ],
        "syncProvidersFromShootSpecDNS": false,
      },
      "type": "shoot-dns-service",
    },
  ],
  "resources": [
    {
      "name": "shoot-dns-service-aws-route53-secret",
      "resourceRef": {
        "apiVersion": "v1",
        "kind": "Secret",
        "name": "aws-route53-secret",
      },
    },
  ],
}
`;

exports[`composables > useShootDns > should add primary dns provider 1`] = `
{
  "dns": {
    "domain": "example.org",
    "providers": [
      {
        "primary": true,
        "secretName": "bar",
        "type": "foo",
      },
    ],
  },
}
`;

exports[`composables > useShootDns > should delete extension dns providers > last provider deleted 1`] = `
{
  "extensions": [],
  "resources": [],
}
`;

exports[`composables > useShootDns > should delete extension dns providers > one provider deleted 1`] = `
{
  "extensions": [
    {
      "providerConfig": {
        "apiVersion": "service.dns.extensions.gardener.cloud/v1alpha1",
        "kind": "DNSConfig",
        "providers": [
          {
            "secretName": undefined,
            "type": "aws-route53",
          },
        ],
        "syncProvidersFromShootSpecDNS": false,
      },
      "type": "shoot-dns-service",
    },
  ],
  "resources": [],
}
`;

exports[`composables > useShootDns > should delete extension dns providers > two providers added 1`] = `
{
  "extensions": [
    {
      "providerConfig": {
        "apiVersion": "service.dns.extensions.gardener.cloud/v1alpha1",
        "kind": "DNSConfig",
        "providers": [
          {
            "secretName": "shoot-dns-service-aws-route53-secret",
            "type": "aws-route53",
          },
          {
            "secretName": undefined,
            "type": "aws-route53",
          },
        ],
        "syncProvidersFromShootSpecDNS": false,
      },
      "type": "shoot-dns-service",
    },
  ],
  "resources": [
    {
      "name": "shoot-dns-service-aws-route53-secret",
      "resourceRef": {
        "apiVersion": "v1",
        "kind": "Secret",
        "name": "aws-route53-secret",
      },
    },
  ],
}
`;
